File: src/main/java/com/avail/descriptor/LiteralPhraseDescriptor.java
/*
 * LiteralNodeDescriptor.java
 * Copyright © 1993-2019, The Avail Foundation, LLC.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of the copyright holder nor the names of the contributors
 *   may be used to endorse or promote products derived from this software
 *   without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package com.avail.descriptor;

import com.avail.annotations.AvailMethod;
import com.avail.compiler.AvailCodeGenerator;
import com.avail.descriptor.PhraseTypeDescriptor.PhraseKind;
import com.avail.descriptor.objects.A_BasicObject;
import com.avail.descriptor.parsing.A_Phrase;
import com.avail.descriptor.parsing.PhraseDescriptor;
import com.avail.descriptor.tuples.A_String;
import com.avail.descriptor.tuples.A_Tuple;
import com.avail.serialization.SerializerOperation;
import com.avail.utility.evaluation.Continuation1NotNull;
import com.avail.utility.evaluation.Transformer1;
import com.avail.utility.json.JSONWriter;

import javax.annotation.Nullable;
import java.util.IdentityHashMap;

import static com.avail.descriptor.AbstractEnumerationTypeDescriptor.instanceTypeOrMetaOn;
import static com.avail.descriptor.LiteralPhraseDescriptor.ObjectSlots.TOKEN;
import static com.avail.descriptor.LiteralTokenDescriptor.literalToken;
import static com.avail.descriptor.LiteralTokenTypeDescriptor.mostGeneralLiteralTokenType;
import static com.avail.descriptor.ObjectTupleDescriptor.tuple;
import static com.avail.descriptor.StringDescriptor.stringFrom;
import static com.avail.descriptor.TokenDescriptor.TokenType.LITERAL;

/**
 * My instances are occurrences of literals parsed from Avail source code.  At
 * the moment only strings and non-negative numbers are supported.
 *
 * @author Mark van Gulik &lt;mark@availlang.org&gt;
 */
public final class LiteralPhraseDescriptor
extends PhraseDescriptor
{
	/**
	 * My slots of type {@link AvailObject}.
	 */
	public enum ObjectSlots
	implements ObjectSlotsEnum
	{
		/**
		 * The token that was transformed into this literal.
		 */
		TOKEN
	}

	@Override
	public void printObjectOnAvoidingIndent (
		final AvailObject object,
		final StringBuilder builder,
		final IdentityHashMap<A_BasicObject, Void> recursionMap,
		final int indent)
	{
		builder.append(object.token().string().asNativeString());
	}

	@Override @AvailMethod
	protected void o_ChildrenDo (
		final AvailObject object,
		final Continuation1NotNull<A_Phrase> action)
	{
		// Do nothing.
	}

	@Override @AvailMethod
	protected void o_ChildrenMap (
		final AvailObject object,
		final Transformer1<A_Phrase, A_Phrase> transformer)
	{
		// Do nothing.
	}

	@Override @AvailMethod
	protected void o_EmitEffectOn (
		final AvailObject object,
		final AvailCodeGenerator codeGenerator)
	{
		// Do nothing.
	}

	@Override @AvailMethod
	protected void o_EmitValueOn (
		final AvailObject object,
		final AvailCodeGenerator codeGenerator)
	{
		codeGenerator.emitPushLiteral(
			tuple(object.token()), object.slot(TOKEN).literal());
	}

	@Override @AvailMethod
	protected boolean o_EqualsPhrase (
		final AvailObject object,
		final A_Phrase aPhrase)
	{
		return !aPhrase.isMacroSubstitutionNode()
			&& object.phraseKind().equals(aPhrase.phraseKind())
			&& object.slot(TOKEN).equals(aPhrase.token());
	}

	@Override @AvailMethod
	protected A_Type o_ExpressionType (final AvailObject object)
	{
		final A_Token token = object.slot(TOKEN);
		assert token.tokenType() == LITERAL;
		final AvailObject literal = token.literal();
		return instanceTypeOrMetaOn(literal).makeImmutable();
	}

	@Override @AvailMethod
	protected int o_Hash (final AvailObject object)
	{
		return object.token().hash() ^ 0x9C860C0D;
	}

	@Override
	protected PhraseKind o_PhraseKind (final AvailObject object)
	{
		return PhraseKind.LITERAL_PHRASE;
	}

	@Override
	protected void o_StatementsDo (
		final AvailObject object,
		final Continuation1NotNull<A_Phrase> continuation)
	{
		throw unsupportedOperationException();
	}

	@Override
	protected SerializerOperation o_SerializerOperation (
		final AvailObject object)
	{
		return SerializerOperation.LITERAL_PHRASE;
	}

	@Override @AvailMethod
	protected A_Token o_Token (final AvailObject object)
	{
		return object.slot(TOKEN);
	}

	@Override
	protected A_Tuple o_Tokens (final AvailObject object)
	{
		return tuple(object.slot(TOKEN));
	}

	@Override @AvailMethod
	protected void o_ValidateLocally (
		final AvailObject object,
		final @Nullable A_Phrase parent)
	{
		// Do nothing.
	}

	@Override
	protected void o_WriteTo (final AvailObject object, final JSONWriter writer)
	{
		writer.startObject();
		writer.write("kind");
		writer.write("literal phrase");
		object.slot(TOKEN).writeTo(writer);
		writer.endObject();
	}

	@Override
	protected void o_WriteSummaryTo (final AvailObject object, final JSONWriter writer)
	{
		writer.startObject();
		writer.write("kind");
		writer.write("literal phrase");
		object.slot(TOKEN).writeSummaryTo(writer);
		writer.endObject();
	}

	/**
	 * Create a literal phrase from a {@linkplain LiteralTokenDescriptor literal
	 * token}.
	 *
	 * @param token The token that describes the literal.
	 * @return The new literal phrase.
	 */
	public static A_Phrase fromTokenForDecompiler (final A_Token token)
	{
		final AvailObject phrase = mutable.create();
		phrase.setSlot(TOKEN, token);
		return phrase.makeShared();
	}

	/**
	 * Create a literal phrase from a {@linkplain LiteralTokenDescriptor literal
	 * token}.
	 *
	 * @param token The token that describes the literal.
	 * @return The new literal phrase.
	 */
	public static A_Phrase literalNodeFromToken (final A_Token token)
	{
		assert token.isInstanceOfKind(mostGeneralLiteralTokenType());
		final AvailObject phrase = mutable.create();
		phrase.setSlot(TOKEN, token);
		return phrase.makeShared();
	}

	/**
	 * Create a literal phrase from an {@link AvailObject}, the literal value
	 * itself.  Automatically wrap the value inside a synthetic literal token.
	 * If the literal is a string, use that as its string form, otherwise use
	 * the {@link StringDescriptor#stringFrom(String)} of its Java
	 * {@link #toString()}.
	 *
	 * @param literalValue
	 *        The value that this literal phrase should produce.
	 * @return The new literal phrase.
	 */
	public static A_Phrase syntheticLiteralNodeFor (
		final A_BasicObject literalValue)
	{
		return syntheticLiteralNodeFor(
			literalValue,
			literalValue.isString()
				? (A_String) literalValue
				: stringFrom(literalValue.toString()));
	}

	/**
	 * Create a literal phrase from an {@link AvailObject}, the literal value
	 * itself.  Automatically wrap the value inside a synthetic literal token.
	 * Use the specified {@code literalAsString} as the string form of the
	 * literal for printing.
	 *
	 * @param literalValue
	 *        The value that this literal phrase should produce.
	 * @param literalAsString
	 *        The {@link A_String} used to describe this literal.
	 * @return The new literal phrase.
	 */
	public static A_Phrase syntheticLiteralNodeFor (
		final A_BasicObject literalValue,
		final A_String literalAsString)
	{
		return literalNodeFromToken(
			literalToken(literalAsString, 0, 0, literalValue));
	}

	/**
	 * Construct a new {@code LiteralPhraseDescriptor}.
	 *
	 * @param mutability
	 *        The {@linkplain Mutability mutability} of the new descriptor.
	 */
	public LiteralPhraseDescriptor (final Mutability mutability)
	{
		super(
			mutability,
			TypeTag.LITERAL_PHRASE_TAG,
			ObjectSlots.class,
			null);
	}

	/** The mutable {@link LiteralPhraseDescriptor}. */
	private static final LiteralPhraseDescriptor mutable =
		new LiteralPhraseDescriptor(Mutability.MUTABLE);

	@Override
	protected LiteralPhraseDescriptor mutable ()
	{
		return mutable;
	}

	/** The shared {@link LiteralPhraseDescriptor}. */
	private static final LiteralPhraseDescriptor shared =
		new LiteralPhraseDescriptor(Mutability.SHARED);

	@Override
	protected LiteralPhraseDescriptor shared ()
	{
		return shared;
	}
}

File: src/main/kotlin/com/avail/descriptor/phrases/LiteralPhraseDescriptor.kt
/*
 * LiteralPhraseDescriptor.kt
 * Copyright © 1993-2020, The Avail Foundation, LLC.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *  * Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 *
 *  * Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 *  * Neither the name of the copyright holder nor the names of the contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package com.avail.descriptor.phrases
import com.avail.compiler.AvailCodeGenerator
import com.avail.descriptor.phrases.A_Phrase.Companion.isMacroSubstitutionNode
import com.avail.descriptor.phrases.A_Phrase.Companion.phraseKind
import com.avail.descriptor.phrases.A_Phrase.Companion.token
import com.avail.descriptor.phrases.LiteralPhraseDescriptor.ObjectSlots.TOKEN
import com.avail.descriptor.representation.A_BasicObject
import com.avail.descriptor.representation.AvailObject
import com.avail.descriptor.representation.Mutability
import com.avail.descriptor.representation.ObjectSlotsEnum
import com.avail.descriptor.tokens.A_Token
import com.avail.descriptor.tokens.LiteralTokenDescriptor
import com.avail.descriptor.tokens.LiteralTokenDescriptor.Companion.literalToken
import com.avail.descriptor.tokens.TokenDescriptor.TokenType
import com.avail.descriptor.tuples.A_String
import com.avail.descriptor.tuples.A_Tuple
import com.avail.descriptor.tuples.ObjectTupleDescriptor.Companion.tuple
import com.avail.descriptor.tuples.ObjectTupleDescriptor.Companion.tupleFromList
import com.avail.descriptor.tuples.StringDescriptor.Companion.stringFrom
import com.avail.descriptor.types.A_Type
import com.avail.descriptor.types.AbstractEnumerationTypeDescriptor.Companion.instanceTypeOrMetaOn
import com.avail.descriptor.types.LiteralTokenTypeDescriptor.Companion.mostGeneralLiteralTokenType
import com.avail.descriptor.types.PhraseTypeDescriptor.PhraseKind
import com.avail.descriptor.types.TypeTag
import com.avail.interpreter.levelOne.L1Decompiler
import com.avail.serialization.SerializerOperation
import com.avail.utility.json.JSONWriter
import java.util.IdentityHashMap

/**
 * My instances are occurrences of literals parsed from Avail source code.  At
 * the moment only strings and non-negative numbers are supported.
 *
 * @constructor
 *
 * @param mutability
 *   The [mutability][Mutability] of the new descriptor.
 *
 * @author Mark van Gulik &lt;mark@availlang.org&gt;
 */
class LiteralPhraseDescriptor(
	mutability: Mutability
) : PhraseDescriptor(
	mutability,
	TypeTag.LITERAL_PHRASE_TAG,
	ObjectSlots::class.java,
	null
) {
	/**
	 * My slots of type [AvailObject].
	 */
	enum class ObjectSlots : ObjectSlotsEnum {
		/**
		 * The token that was transformed into this literal.
		 */
		TOKEN
	}

	override fun printObjectOnAvoidingIndent(
		self: AvailObject,
		builder: StringBuilder,
		recursionMap: IdentityHashMap<A_BasicObject, Void>,
		indent: Int
	) {
		builder.append(self.token().string().asNativeString())
	}

	override fun o_ChildrenDo(self: AvailObject, action: (A_Phrase) -> Unit) {
		// Do nothing.
	}

	override fun o_ChildrenMap(
		self: AvailObject,
		transformer: (A_Phrase) -> A_Phrase
	) {
		// Do nothing.
	}

	override fun o_EmitEffectOn(
		self: AvailObject,
		codeGenerator: AvailCodeGenerator
	) {
		// Do nothing.
	}

	override fun o_EmitValueOn(
		self: AvailObject,
		codeGenerator: AvailCodeGenerator
	) = codeGenerator.emitPushLiteral(
		tuple(self.token()), self.slot(TOKEN).literal())

	override fun o_EqualsPhrase(
		self: AvailObject,
		aPhrase: A_Phrase
	) = (!aPhrase.isMacroSubstitutionNode()
		&& self.phraseKind() == aPhrase.phraseKind()
		&& self.slot(TOKEN).equals(aPhrase.token()))

	/**
	 * Simplify decompilation by pretending a literal phrase holding tuple is
	 * actually a list phrase of literal phrases for each element.
	 *
	 * This only comes into play if the [L1Decompiler] was about to fail anyhow.
	 */
	override fun o_ExpressionsTuple(self: AvailObject): A_Tuple =
		tupleFromList(self.map { syntheticLiteralNodeFor(it) })

	override fun o_PhraseExpressionType(self: AvailObject): A_Type {
		val token: A_Token = self.slot(TOKEN)
		assert(token.tokenType() === TokenType.LITERAL)
		return instanceTypeOrMetaOn(token.literal()).makeImmutable()
	}

	override fun o_Hash(self: AvailObject): Int =
		self.token().hash() xor -0x6379f3f3

	override fun o_PhraseKind(self: AvailObject): PhraseKind =
		PhraseKind.LITERAL_PHRASE

	override fun o_StatementsDo(
		self: AvailObject,
		continuation: (A_Phrase) -> Unit
	): Unit = unsupported

	override fun o_SerializerOperation(self: AvailObject): SerializerOperation =
		SerializerOperation.LITERAL_PHRASE

	override fun o_Token(self: AvailObject): A_Token = self.slot(TOKEN)

	override fun o_Tokens(self: AvailObject): A_Tuple = tuple(self.slot(TOKEN))

	override fun o_ValidateLocally(
		self: AvailObject,
		parent: A_Phrase?
	) {
		// Do nothing.
	}

	override fun o_WriteTo(self: AvailObject, writer: JSONWriter) =
		writer.writeObject {
			at("kind") { write("literal phrase") }
			at("token") { self.slot(TOKEN).writeTo(writer) }
		}

	override fun o_WriteSummaryTo(self: AvailObject, writer: JSONWriter) =
		writer.writeObject {
			at("kind") { write("literal phrase") }
			at("token") { self.slot(TOKEN).writeSummaryTo(writer) }
		}

	override fun mutable() = mutable

	override fun shared() = shared

	companion object {
		/**
		 * Create a literal phrase from a [literal][LiteralTokenDescriptor].
		 *
		 * @param token
		 *   The token that describes the literal.
		 * @return
		 *   The new literal phrase.
		 */
		fun fromTokenForDecompiler(token: A_Token): A_Phrase =
			mutable.createShared {
				setSlot(TOKEN, token)
			}

		/**
		 * Create a literal phrase from a [literal][LiteralTokenDescriptor].
		 *
		 * @param token
		 *   The token that describes the literal.
		 * @return
		 *   The new literal phrase.
		 */
		fun literalNodeFromToken(token: A_Token): A_Phrase {
			assert(token.isInstanceOfKind(mostGeneralLiteralTokenType()))
			return mutable.createShared {
				setSlot(TOKEN, token)
			}
		}

		/**
		 * Create a literal phrase from an [AvailObject], the literal value
		 * itself.  Automatically wrap the value inside a synthetic literal
		 * token. Use the specified `literalAsString` as the string form of the
		 * literal for printing.
		 *
		 * @param literalValue
		 *   The value that this literal phrase should produce.
		 * @param literalAsString
		 *   The optional [A_String] used to describe this literal.
		 * @return
		 *   The new literal phrase.
		 */
		@JvmOverloads
		fun syntheticLiteralNodeFor(
			literalValue: A_BasicObject,
			literalAsString: A_String =
				if (literalValue.isString) literalValue as A_String
				else stringFrom(literalValue.toString())
		): A_Phrase = literalNodeFromToken(
			literalToken(literalAsString, 0, 0, literalValue))

		/** The mutable [LiteralPhraseDescriptor].  */
		private val mutable = LiteralPhraseDescriptor(Mutability.MUTABLE)

		/** The shared [LiteralPhraseDescriptor].  */
		private val shared = LiteralPhraseDescriptor(Mutability.SHARED)
	}
}
